Source/WebCore/crypto/mac/CryptoKeyECMac.cpp

```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+    cced25519_make_key_pair(di, rng, ccPublicKey, ccPrivateKey);
+    auto publicKey = CryptoKeyEC::create(identifier, curve, CryptoKeyType::Public, Vector<uint8_t>(ccPublicKey), true, usages);
+    auto privateKey = CryptoKeyEC::create(identifier, curve, CryptoKeyType::Private, Vector<uint8_t>(ccPrivateKey), extractable, usages);
     return CryptoKeyPair { WTFMove(publicKey), WTFMove(privateKey) };
+    
 }
 
 RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportRaw(CryptoAlgorithmIdentifier identifier, NamedCurve curve, Vector<uint8_t>&& keyData, bool extractable, CryptoKeyUsageBitmap usages)
 {
-    if (!doesUncompressedPointMatchNamedCurve(curve, keyData.size()))
-        return nullptr;
-
-    CCECCryptorRef ccPublicKey = nullptr;
-    if (CCECCryptorImportKey(kCCImportKeyBinary, keyData.data(), keyData.size(), ccECKeyPublic, &ccPublicKey))
+    if (curve != NamedCurve::Curve25519) {
+        if (!doesUncompressedPointMatchNamedCurve(curve, keyData.size()))
+            return nullptr;
+        CCECCryptorRef ccPublicKey = nullptr;
+        if (CCECCryptorImportKey(kCCImportKeyBinary, keyData.data(), keyData.size(), ccECKeyPublic, &ccPublicKey))
+            return nullptr;
+        return create(identifier, curve, CryptoKeyType::Public, std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>(ccPublicKey), extractable, usages);
+        
+    }
+    if (!doesFieldElementMatchNamedCurve(curve, keyData.size()))
         return nullptr;
-
-    return create(identifier, curve, CryptoKeyType::Public, PlatformECKeyContainer(ccPublicKey), extractable, usages);
+    if (usages & CryptoKeyUsageSign)
+        return create(identifier, curve, CryptoKeyType::Private, Vector<uint8_t>(keyData), extractable, usages);
+    return create(identifier, curve, CryptoKeyType::Public, Vector<uint8_t>(keyData), extractable, usages);
+        
 }
 
 Vector<uint8_t> CryptoKeyEC::platformExportRaw() const
 {
-    size_t expectedSize = 2 * keySizeInBytes() + 1; // Per Section 2.3.4 of http://www.secg.org/sec1-v2.pdf
-    Vector<uint8_t> result(expectedSize);
-    size_t size = result.size();
-    if (UNLIKELY(CCECCryptorExportKey(kCCImportKeyBinary, result.data(), &size, ccECKeyPublic, m_platformKey.get()) || size != expectedSize))
-        return { };
-    return result;
+    // Per Section 2.3.4 of http://www.secg.org/sec1-v2.pdf
+    return WTF::switchOn( m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) -> Vector<uint8_t> {
+        size_t expectedSize = 2 * keySizeInBytes() + 1;
+        Vector<uint8_t> result(expectedSize);
+        size_t size = result.size();
+        if (UNLIKELY(CCECCryptorExportKey(kCCImportKeyBinary, result.data(), &size, ccECKeyPublic, options.get()) || size != expectedSize))
+            return { };
+        return result;
+        },
+        [&](const Vector<uint8_t>& options) -> Vector<uint8_t> {
```

Comment:
The content of `options` isn't used at all here, You are just returning a result vector with `expectedSize` items that are all 0.

Reply:
changed